From 71a7ce097bfea33bb7096c30e282f97e334ebda7 Mon Sep 17 00:00:00 2001
From: peak3d <pfau@peak3d.de>
Date: Wed, 2 Aug 2017 11:04:10 +0200
Subject: [PATCH 02/16] [HLS] int64 PTS istead uint64 / support
 #EXT-X-STREAM-INF without codecs / no streamchange if fps changes

---
 src/elementaryStream.h | 12 ++++++------
 src/tsDemuxer.cpp      |  8 +++++---
 2 files changed, 11 insertions(+), 9 deletions(-)

diff --git a/src/elementaryStream.h b/src/elementaryStream.h
index 3064515..6a49f37 100644
--- a/src/elementaryStream.h
+++ b/src/elementaryStream.h
@@ -82,8 +82,8 @@ namespace TSDemux
     uint16_t              pid;
     size_t                size;
     const unsigned char*  data;
-    uint64_t              dts;
-    uint64_t              pts;
+    int64_t               dts;
+    int64_t               pts;
     uint64_t              duration;
     bool                  streamChange;
   };
@@ -101,10 +101,10 @@ namespace TSDemux
 
     uint16_t pid;
     STREAM_TYPE stream_type;
-    uint64_t c_dts;               ///< current MPEG stream DTS (decode time for video)
-    uint64_t c_pts;               ///< current MPEG stream PTS (presentation time for audio and video)
-    uint64_t p_dts;               ///< previous MPEG stream DTS (decode time for video)
-    uint64_t p_pts;               ///< previous MPEG stream PTS (presentation time for audio and video)
+    int64_t c_dts;               ///< current MPEG stream DTS (decode time for video)
+    int64_t c_pts;               ///< current MPEG stream PTS (presentation time for audio and video)
+    int64_t p_dts;               ///< previous MPEG stream DTS (decode time for video)
+    int64_t p_pts;               ///< previous MPEG stream PTS (presentation time for audio and video)
 
     bool has_stream_info;         ///< true if stream info is completed else it requires parsing of iframe
 
diff --git a/src/tsDemuxer.cpp b/src/tsDemuxer.cpp
index 578fc11..ac29636 100644
--- a/src/tsDemuxer.cpp
+++ b/src/tsDemuxer.cpp
@@ -1040,7 +1040,7 @@ int AVContext::parse_ts_pes()
         has_pts = true;
         if (this->packet->packet_table.len >= 14)
         {
-          uint64_t pts = decode_pts(this->packet->packet_table.buf + 9);
+          int64_t pts = decode_pts(this->packet->packet_table.buf + 9);
           this->packet->stream->p_dts = this->packet->stream->c_dts;
           this->packet->stream->p_pts = this->packet->stream->c_pts;
           this->packet->stream->c_dts = this->packet->stream->c_pts = pts;
@@ -1056,8 +1056,10 @@ int AVContext::parse_ts_pes()
         has_pts = true;
         if (this->packet->packet_table.len >= 19 )
         {
-          uint64_t pts = decode_pts(this->packet->packet_table.buf + 9);
-          uint64_t dts = decode_pts(this->packet->packet_table.buf + 14);
+          int64_t pts = decode_pts(this->packet->packet_table.buf + 9);
+          int64_t dts = decode_pts(this->packet->packet_table.buf + 14);
+          if (pts < dts)
+            dts -= PTS_UNSET;
           int64_t d = (pts - dts) & PTS_MASK;
           // more than two seconds of PTS/DTS delta, probably corrupt
           if(d > 180000)
-- 
2.17.1.windows.2

